db/crud.go

```diff
@@ -814,6 +814,240 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string,
+		// on the feed is small - sub-second, so we usually shouldn't care about more than
+		// a few recent sequences.  However, the pruning has some overhead (read lock on nextSequence),
+		// so we're allowing more 'recent sequences' on the doc (20) before attempting pruning
+		stableSequence := db.changeCache.GetStableSequence(docid).Seq
+		count := 0
+		for _, seq := range doc.RecentSequences {
+			// Only remove sequences if they are higher than a sequence that's been seen on the
+			// feed. This is valid across SG nodes (which could each have a different nextSequence),
+			// as the mutations that this node used to rev nextSequence will at some point be delivered
+			// to each node.
+			if seq < stableSequence {
+				count++
+			} else {
+				break
+			}
+		}
+		if count > 0 {
+			doc.RecentSequences = doc.RecentSequences[count:]
+		}
+	}
+
+	// Append current sequence and unused sequences to recent sequence history
+	doc.RecentSequences = append(doc.RecentSequences, unusedSequences...)
+	doc.RecentSequences = append(doc.RecentSequences, docSequence)
+
+	return unusedSequences, docSequence, nil
+}
+
+// Bool is for bail out
+func (db *Database) updateRevTree(doc *Document, prevCurrentRev, newRevID, docid string, changedPrincipals []string,
+	changedUsers []string, access channels.AccessMap, roles channels.AccessMap, channelSet base.Set, syncExpiry *uint32,
+	oldBody string) (
+	retChannelSet base.Set, retAccess channels.AccessMap, retRoles channels.AccessMap, retSyncExpiry *uint32,
+	retOldBodyJSON string, retChangedPrincipals []string, retChangedRoleUsers []string, retReturnEarly bool,
+	retErr error) {
+	var err error
+	if doc.CurrentRev != prevCurrentRev {
+		// Most of the time this update will change the doc's current rev. (The exception is
+		// if the new rev is a conflict that doesn't win the revid comparison.) If so, we
+		// need to update the doc's top-level Channels and Access properties to correspond
+		// to the current rev's state.
+		if newRevID != doc.CurrentRev {
+			// In some cases an older revision might become the current one. If so, get its
+			// channels & access, for purposes of updating the doc:
+			var curBody Body
+			if curBody, err = db.getAvailableRev(doc, doc.CurrentRev); curBody != nil {
+				base.DebugfCtx(db.Ctx, base.KeyCRUD, "updateDoc(%q): Rev %q causes %q to become current again",
+					base.UD(docid), newRevID, doc.CurrentRev)
+				channelSet, access, roles, syncExpiry, oldBody, err = db.getChannelsAndAccess(doc, curBody, doc.CurrentRev)
+				if err != nil {
+					return channelSet, access, roles, syncExpiry, oldBody, changedPrincipals, changedUsers, false, err
+				}
+			} else {
+				// Shouldn't be possible (CurrentRev is a leaf so won't have been compacted)
+				base.WarnfCtx(db.Ctx, base.KeyAll, "updateDoc(%q): Rev %q missing, can't call getChannelsAndAccess "+
+					"on it (err=%v)", base.UD(docid), doc.CurrentRev, err)
```

Comment:
We've lost this from here, and it seems like a valid thing to still be doing?

```
channelSet = nil
access = nil
roles = nil
```

Reply:
Yeah you're right. An earlier version I had basically re-created those vars so would've been null at this point but yeah...